src/client/src/database.rs

```diff
@@ -283,55 +288,115 @@ impl Database {
-            .collect::<Result<Vec<_>>>()?;
-
-        let output = if let Some(FlightMessage::AffectedRows(rows)) = flight_messages.get(0) {
-            ensure!(
-                flight_messages.len() == 1,
-                IllegalFlightMessagesSnafu {
-                    reason: "Expect 'AffectedRows' Flight messages to be one and only!"
-                }
-            );
-            Output::AffectedRows(*rows)
-        } else {
-            let recordbatches = flight_messages_to_recordbatches(flight_messages)
-                .context(ConvertFlightDataSnafu)?;
-            Output::RecordBatches(recordbatches)
+                    );
+                    error
+                })
+                .unwrap_err()
+        })?;
+
+        let flight_data_stream = response.into_inner();
+        let mut decoder = FlightDecoder::default();
+
+        let mut flight_message_stream = flight_data_stream.map_ok(move |flight_data| {
+            decoder
+                .try_decode(flight_data)
+                .context(ConvertFlightDataSnafu)
+        });
+
+        let Some(first_flight_message) = flight_message_stream.next().await else {
+            return Ok(Output::RecordBatches(RecordBatches::empty()));
         };
-        Ok(output)
+
+        match first_flight_message {
+            // tonic error.
+            Err(status) => {
+                let e: error::Error = status.into();
+                let code = e.status_code();
+                let msg = e.to_string();
+                ServerSnafu { code, msg }.fail()
+            }
+            // decode error.
+            Ok(Err(err)) => Err(err),
+            Ok(Ok(first_flight_message_raw)) => match first_flight_message_raw {
+                FlightMessage::AffectedRows(rows) => {
+                    ensure!(
+                        flight_message_stream.next().await.is_none(),
+                        IllegalFlightMessagesSnafu {
+                            reason:
+                                "Expect 'AffectedRows' Flight messages to be the one and the only!"
+                        }
+                    );
+                    Ok(Output::AffectedRows(rows))
+                }
+                FlightMessage::Schema(schema) => {
+                    let stream: Pin<Box<dyn Stream<Item = RecordBatchResult<RecordBatch>> + Send>> =
+                        Box::pin(stream! {
+                            while let Some(flight_message) = flight_message_stream.next().await {
+                                match flight_message {
+                                    // tonic error.
+                                    Err(status) => {
+                                        let e: error::Error = status.into();
+                                        let code = e.status_code();
+                                        let msg = e.to_string();
+                                        yield ServerSnafu { code, msg }
```

Comment:
It seems we already have `impl From<Status> for Error` in client error mod, can we reuse that?